legacy/c3_trajectory_generator/src/node.cpp

```diff
@@ -167,21 +181,51 @@ struct Node {
     current_waypoint_t = now;
   }
 
-  void timer_callback(const ros::TimerEvent &) {
-    if (!c3trajectory) return;
+  void timer_callback(const ros::TimerEvent &)
+  {
+    if (!c3trajectory)
+      return;
 
     ros::Time now = ros::Time::now();
 
-    if (actionserver.isNewGoalAvailable()) {
+    auto old_waypoint = current_waypoint;
+
+    if (actionserver.isNewGoalAvailable())
+    {
       boost::shared_ptr<const mil_msgs::MoveToGoal> goal = actionserver.acceptNewGoal();
       current_waypoint = subjugator::C3Trajectory::Waypoint(
-          Point_from_PoseTwist(goal->posetwist.pose, goal->posetwist.twist), goal->speed,
-          !goal->uncoordinated);
+          Point_from_PoseTwist(goal->posetwist.pose, goal->posetwist.twist), goal->speed, !goal->uncoordinated);
       current_waypoint_t = now;
       this->linear_tolerance = goal->linear_tolerance;
       this->angular_tolerance = goal->angular_tolerance;
+
+      // Check if waypoint is valid
+      std::pair<bool, WAYPOINT_ERROR_TYPE> checkWPResult =
+          waypoint_validity_.is_waypoint_valid(Pose_from_Waypoint(current_waypoint));
+      if (checkWPResult.first == false)  // got a point that we should not move to
```

Comment:
You can set it to the water surface for now, and then we'll see if we want to change it to something different after testing.

You always want to make sure that your'e handling all of the possibilities when you start branching off with `if` statements.